src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -83,56 +83,157 @@ ShenandoahMmuTracker::~ShenandoahMmuTracker() {
+
+size_t ShenandoahGenerationSizer::calculate_min_size(size_t heap_size) {
+  size_t default_value = (heap_size * ShenandoahMinYoungPercentage) / 100;
+  return MAX2(ShenandoahHeapRegion::region_size_bytes(), default_value);
+}
+
+size_t ShenandoahGenerationSizer::calculate_max_size(size_t heap_size) {
+  size_t default_value = (heap_size * ShenandoahMaxYoungPercentage) / 100;
+  return MAX2(ShenandoahHeapRegion::region_size_bytes(), default_value);
+}
+
+void ShenandoahGenerationSizer::recalculate_min_max_young_length(size_t heap_size) {
+  assert(heap_size > 0, "Heap must be initialized");
+
+  switch (_sizer_kind) {
+    case SizerDefaults:
+      _min_desired_young_size = calculate_min_size(heap_size);
+      _max_desired_young_size = calculate_max_size(heap_size);
+      break;
+    case SizerNewSizeOnly:
+      _max_desired_young_size = calculate_max_size(heap_size);
+      _max_desired_young_size = MAX2(_min_desired_young_size, _max_desired_young_size);
+      break;
+    case SizerMaxNewSizeOnly:
+      _min_desired_young_size = calculate_min_size(heap_size);
+      _min_desired_young_size = MIN2(_min_desired_young_size, _max_desired_young_size);
+      break;
+    case SizerMaxAndNewSize:
+      // Do nothing. Values set on the command line, don't update them at runtime.
+      break;
+    case SizerNewRatio:
+      _min_desired_young_size = MAX2((heap_size / (NewRatio + 1)), ShenandoahHeapRegion::region_size_bytes());
+      _max_desired_young_size = _min_desired_young_size;
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+
+  assert(_min_desired_young_size <= _max_desired_young_size, "Invalid min/max young gen size values");
+}
+
+void ShenandoahGenerationSizer::heap_size_changed(size_t heap_size) {
+  recalculate_min_max_young_length(heap_size);
+}
+
+bool ShenandoahGenerationSizer::adjust_generation_sizes() {
   shenandoah_assert_generational();
-  if (_mmu_average.davg() >= double(GCTimeRatio)) {
+  if (!use_adaptive_sizing()) {
+    return false;
+  }
+
+  if (_mmu_tracker->average() >= double(GCTimeRatio)) {
     return false;
   }
 
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   ShenandoahOldGeneration *old = heap->old_generation();
-  double old_time_s = old->reset_collection_time();
   ShenandoahYoungGeneration *young = heap->young_generation();
-  double young_time_s = young->reset_collection_time();
   ShenandoahGeneration *global = heap->global_generation();
+  double old_time_s = old->reset_collection_time();
+  double young_time_s = young->reset_collection_time();
   double global_time_s = global->reset_collection_time();
 
+  const double transfer_threshold = 3.0;
+  double delta = young_time_s - old_time_s;
+
   log_info(gc)("Thread Usr+Sys YOUNG = %.3f, OLD = %.3f, GLOBAL = %.3f", young_time_s, old_time_s, global_time_s);
 
-  if (old_time_s > young_time_s) {
-    return transfer_capacity(young, old);
-  } else {
-    return transfer_capacity(old, young);
+  if (abs(delta) <= transfer_threshold) {
```

Comment:
I thought the original idea was to use the difference in MMU's for old and young as the error signal to drive the (direction of the) transfer, rather than the difference in the actual times? Am I misinterpreting what `reset_collection_time` returns? You do refer to it as `thread utilization` (akin to MMU) in the log message below.

Reply:
I reasoned that the denominator in a comparison between MMU's (i.e., process time) would be the same on both sides of the comparison, so I omitted it.

Reply:
My worry is that using the latest sample might lead to too much variance. Using a TruncatedSeq's decaying average would help with damping the error signal and lead to smoother control over sizing changes.

Reply:
I'll experiment with that. I was going for responsiveness. Smoothing out the signal may also introduce a delay in resizing.